src/shortest_path/all_pairs_bellman_ford.rs

```diff
@@ -0,0 +1,202 @@
+            let path_lenghts: Option<Vec<Option<f64>>> =
+                bellman_ford(graph, x, |e| edge_cost(e.id()), None).unwrap();
+
+            if path_lenghts.is_none() {
+                let mut cycle = negative_cycle.write().unwrap();
+                *cycle = true;
+                return (
+                    x.index(),
+                    PathLengthMapping {
+                        path_lengths: DictMap::new(),
+                    },
+                );
+            }
+
+            let out_map = PathLengthMapping {
+                path_lengths: path_lenghts
+                    .unwrap()
+                    .into_iter()
+                    .enumerate()
+                    .filter_map(|(index, opt_cost)| {
+                        if index != x.index() {
+                            opt_cost.map(|cost| (index, cost))
+                        } else {
+                            None
+                        }
+                    })
+                    .collect(),
+            };
+            (x.index(), out_map)
+        })
+        .collect();
+
+    if *negative_cycle.read().unwrap() {
+        return Err(NegativeCycle::new_err(
+            "The shortest-path is not defined because there is a negative cycle",
+        ));
+    }
+
+    Ok(AllPairsPathLengthMapping {
+        path_lengths: out_map,
+    })
+}
+
+pub fn all_pairs_bellman_ford_shortest_paths<Ty: EdgeType + Sync>(
+    py: Python,
+    graph: &StablePyGraph<Ty>,
+    edge_cost_fn: PyObject,
+) -> PyResult<AllPairsPathMapping> {
+    if graph.node_count() == 0 {
+        return Ok(AllPairsPathMapping {
+            paths: DictMap::new(),
+        });
+    } else if graph.edge_count() == 0 {
+        return Ok(AllPairsPathMapping {
+            paths: graph
+                .node_indices()
+                .map(|i| {
+                    (
+                        i.index(),
+                        PathMapping {
+                            paths: DictMap::new(),
+                        },
+                    )
+                })
+                .collect(),
+        });
+    }
+    let edge_weights: Vec<Option<f64>> =
+        edge_weights_from_callable(py, graph, &Some(edge_cost_fn), 1.0)?;
+    let edge_cost = |e: EdgeIndex| -> PyResult<f64> {
+        match edge_weights[e.index()] {
+            Some(weight) => Ok(weight),
+            None => Err(PyIndexError::new_err("No edge found for index")),
+        }
+    };
+
+    let node_indices: Vec<NodeIndex> = graph.node_indices().collect();
+
+    let negative_cycle = RwLock::new(false);
+
+    let out_map = AllPairsPathMapping {
+        paths: node_indices
+            .into_par_iter()
+            .map(|x| {
+                let mut paths: DictMap<NodeIndex, Vec<NodeIndex>> =
+                    DictMap::with_capacity(graph.node_count());
+                let path_lenghts: Option<Vec<Option<f64>>> =
+                    bellman_ford(graph, x, |e| edge_cost(e.id()), Some(&mut paths)).unwrap();
```

Comment:
Same comment as above

```suggestion
                if negative_cycle.read().unwrap() {
                    return (
                        x.index(),
                        PathMapping {
                            paths: DictMap::new(),
                        },
                    );
                }
                let mut paths: DictMap<NodeIndex, Vec<NodeIndex>> =
                    DictMap::with_capacity(graph.node_count());
                let path_lenghts: Option<Vec<Option<f64>>> =
                    bellman_ford(graph, x, |e| edge_cost(e.id()), Some(&mut paths)).unwrap();
```

Reply:
Done